### PR TITLE
fix(providers): drop stale signed thinking blocks after a mid-conversation model switch

### PR DESCRIPTION
**Pull request by kyledef**
## Problem (root cause)

Extended-thinking agent sessions die with an unrecoverable provider 400:

```
400 BAD_REQUEST — messages.N.content.M: `thinking` or `redacted_thinking` blocks in the
latest assistant message cannot be modified. These blocks must remain as they were in the
original response.
```

A thinking block's signature is issued by — and only valid for — the model that produced it. When a conversation switches models or thinking-effort **mid-stream** (e.g. via `set_config_option`), the stored history still carries the *previous* model's signed `thinking` / `redacted_thinking` blocks (Databricks calls them `reasoning` blocks). Replaying those signatures against a different model makes Anthropic — direct and via Databricks/Bedrock — reject the whole request. The agent turn then ends unexpectedly.

This was observed in the wild: two `set_config_option` calls, then an immediate 400 on the next request (at message index 3, so not a compaction artifact).

## Fix

When serializing history, drop signed thinking / redacted-thinking blocks whose **originating model** (from `message.metadata.inference`) differs from the model the current request targets. The turn's text and tool content are still sent.

When provenance is unknown — older stored messages, or no inference metadata — the prior behavior is kept, so **single-model conversations are completely unaffected** (no regression).

Applied in both serialization paths:
- **Anthropic formatter** (`formats/anthropic.rs`): new `AnthropicFormatOptions.current_model` + shared `thinking_block_is_stale()`; skips stale signed `thinking` and all stale `redacted_thinking`.
- **Databricks formatter** (`formats/databricks.rs`) — the path in the observed incident: same staleness check applied to its `reasoning` summary blocks, threading the target model into `format_messages`.

## Tests

For both formatters:
- stale block from a *different* model is dropped, text content preserved;
- block from the *same* model is kept (signature intact);
- unknown provenance is kept.

All 140 `providers::formats` tests pass; `cargo fmt` and clippy clean (no new warnings in touched files).

## Related

Pairs with the retry-side fix (#10005), which stops the identical 400 payload from being retried 3×. This PR removes the cause; that PR makes any remaining permanent 400 fail fast instead of after wasteful retries.

Internal tracking: BOT-1019.

### CODE REVIEW

**Comment by kyledef** [comment]
Thanks @codex — both P2 findings were valid and are now addressed in `8313966`:

1. **Databricks endpoint vs. resolved upstream name**: `thinking_block_is_stale` now compares against `requested_model` (the endpoint identity space callers pass as `current_model`) instead of `resolved_model` (which holds the upstream name for Databricks). Added a regression test `keeps_reasoning_when_endpoint_matches_despite_upstream_resolved_name`.

2. **Missing provenance for direct Anthropic**: `Agent::reply_internal` no longer gates `InferenceMetadata` on `resolved_model` being present — it always records at least `requested_model`, so direct Anthropic extended-thinking sessions now carry the provenance the formatter needs.

141 `providers::formats` tests pass; fmt + clippy clean.

File: crates/goose-provider-types/src/formats/databricks.rs [hunk 1]
```diff
@@ -641,7 +654,7 @@ pub fn create_request_for_provider(
         tool_call_id: None,
     };
 
-    let messages_spec = format_messages(messages, image_format);
+    let messages_spec = format_messages(messages, image_format, Some(&model_config.model_name));
```
[the remark this answers is not in the document]
Comment (kyledef) [reply crates/goose-provider-types/src/formats/databricks.rs:535]: Addressed: the check now compares against `requested_model` (the endpoint-name identity space callers pass as `current_model`) rather than `resolved_model`, so a same-endpoint Databricks follow-up is no longer flagged stale despite its upstream name differing. Added regression test `keeps_reasoning_when_endpoint_matches_despite_upstream_resolved_name` to lock this in.

File: crates/goose-provider-types/src/formats/anthropic.rs [hunk 2]
```diff
@@ -76,10 +81,35 @@ impl AnthropicFormatOptions {
...
+/// model-specific signatures that Anthropic rejects if replayed against a
+/// different model. Returns false when we can't establish provenance (no
+/// `current_model`, or the message has no recorded originating model), so
+/// single-model conversations are unaffected.
+pub fn thinking_block_is_stale(message: &Message, current_model: Option<&str>) -> bool {
+    let Some(current_model) = current_model else {
+        return false;
+    };
+    let Some(inference) = message.metadata.inference.as_ref() else {
+        return false;
```
[the remark this answers is not in the document]
Comment (kyledef) [reply crates/goose-provider-types/src/formats/anthropic.rs:85]: Addressed: `Agent::reply_internal` no longer gates `InferenceMetadata` on `resolved_model` being present — it now always records at least `requested_model`, so direct Anthropic extended-thinking sessions carry the provenance this formatter needs. The "no inference metadata" early-return now only affects genuinely pre-existing/legacy messages.

File: crates/goose/src/providers/formats/anthropic.rs [hunk 3]
```diff
@@ -76,10 +81,36 @@ impl AnthropicFormatOptions {
...
+        return false;
+    };
+    // Compare in the same identity space as `current_model`, which callers pass
+    // as the model/endpoint name from `ModelConfig` (`requested_model`). Avoid
+    // `resolved_model`: for Databricks that holds the *upstream* model name
+    // (e.g. "claude-opus-4.6") while `current_model` is the endpoint name (e.g.
+    // "databricks-claude-opus-4-6"), so comparing them would flag valid blocks
+    // as stale on a same-endpoint follow-up.
+    let origin_model = inference.requested_model.as_str();
+    !origin_model.is_empty() && origin_model != current_model
```
[the remark this answers is not in the document]
Comment (kyledef) [reply crates/goose/src/providers/formats/anthropic.rs]: This is the direct counterpoint to the earlier P2 on line 657, and the two can't both be satisfied with one comparison: comparing in resolved-model space fixes the rare repointed-endpoint case but reintroduces false-positive stripping for every Databricks endpoint whose name differs from its upstream, which is the common case. I opted for the endpoint-name (`requested_model`) comparison as the safe default and rely on the permanent-400 retry guard in #10005 as the backstop for the repointed-endpoint edge case; happy to switch to resolved-space if you'd prefer that tradeoff.

**Comment by kyledef** [comment]
Rebased onto latest `main` (the provider crate split moved the Anthropic format into `goose-providers`; logic and tests carried over, build/clippy clean). While re-reviewing I also hardened `thinking_block_is_stale` to match against *both* `requested_model` (endpoint identity) and `resolved_model` (upstream identity), so a block is only stale when it matches neither — this closes the remaining edge case where a Databricks Claude-backed endpoint rewrites the format config to the upstream name. Added regression test `keeps_reasoning_when_current_model_matches_upstream_resolved_name`; 55 databricks + 45 anthropic format tests pass.

**Comment by DOsinga** [comment]
Thanks @kyledef — the fix itself looks right and the codex threads are well handled. Two cleanups before we land it:

**Comments** — the model-signature/400 rationale is currently explained in several places: the doc comment on `current_model`, the doc comment on `thinking_block_is_stale()`, and the block comment inside `format_messages_with_options`. Keep the "why" in one canonical place (the `thinking_block_is_stale()` doc comment is the natural home) and drop the rest. Please also remove the inline restatements that just narrate the guard they sit above:
- `// Signed blocks are model-specific; skip if stale.`
- `// Redacted thinking is always signed/model-specific; skip if stale.`
- the per-assert comments in the tests (`// Signed thinking block dropped...`, `// No inference metadata recorded -> ...`, `// Same model: ...` / `// Different model: ...`) — the assert messages already say this.

**Tests** — `drops_stale_reasoning_blocks_from_a_different_model` actually tests two things (same-model kept *and* different-model dropped). Please split it into two tests like you did on the anthropic side so the names match what they assert. Also, the `match` that accepts the text content as *either* a bare `String` *or* an `Array` is hedging — you know the exact output shape here, so assert that one shape rather than passing on both.

**Comment by kyledef** [comment]
Rebased onto the current `main` (provider-crate split moved `formats/anthropic.rs`/`formats/databricks.rs` into `goose-provider-types`, and `AnthropicProvider::from_custom_config` now delegates to `goose_providers::anthropic::from_declarative_config`; resolved both conflicts, logic carried over, 217 `goose-provider-types::formats` tests + 14 `agents::agent` tests pass, fmt/clippy clean).

Addressed both review-comment cleanups from @DOsinga in `b1355a674`:
- Consolidated the model-signature/400 rationale into the `thinking_block_is_stale()` doc comment (the canonical home); dropped the duplicated blurb on `current_model` and the block comment inside `format_messages_with_options`, plus the narrating inline comments at each guard.
- Split `drops_stale_reasoning_blocks_from_a_different_model` into `keeps_reasoning_block_from_the_same_model` / `drops_reasoning_block_from_a_different_model` so names match what they assert, matching the anthropic-side pattern. Replaced the `String`-or-`Array` hedge with a direct `assert_eq!` on the bare-string shape (confirmed that's the only shape possible once reasoning is dropped and only text remains).

**Comment by kyledef** [comment]
Ran an automated review pass (`sq agents review`) after the rebase — it correctly recognized the two earlier Codex threads as already resolved. One new, non-blocking finding worth tracking:

`thinking_block_is_stale` compares model identity (`requested_model`/`resolved_model`) only. If a user changes `thinking_effort` on the *same* model (`Agent::update_thinking_effort`), model identity is unchanged, so the check treats the prior signed block as fresh and replays it — potentially the same 400 this PR fixes, just triggered by an effort change instead of a model change. The original problem statement above does call out "switches models **or thinking-effort** mid-stream," so this may be a real residual gap in the same bug class.

Not fixing it in this PR: I haven't confirmed live whether Anthropic's signature check is actually keyed to effort/budget_tokens vs. purely model identity, this PR already resolves the reported incident, and #10005's retry backstop still fails fast on any residual 400 of this shape rather than retrying 3x. Suggest a fast follow-up PR that records the originating `thinking_effort` in `InferenceMetadata` and folds it into the staleness comparison, mirroring the model-identity pattern here. Happy to pick that up if useful.

**Review by DOsinga — APPROVED** [review]
